app/models/gem_download.rb

```diff
@@ -44,24 +44,41 @@ def self.update_count_by(count, rubygem_id:, version_id: 0)
     find_by_sql([update, count]).first
   end
 
-  def self.increment(full_name, count: 1)
-    version = Version.find_by(full_name: full_name)
-    return unless version
-    # Total count
-    update_count_by(count, rubygem_id: 0, version_id: 0)
-    # Gem count
-    update_count_by(count, rubygem_id: version.rubygem_id, version_id: 0)
-    # Gem version count
-    update_count_by(count, rubygem_id: version.rubygem_id, version_id: version.id)
-  end
-
   # Takes an array where members have the form
   #   [full_name, count]
   # E.g.:
   #   ['rake-10.4.2', 1]
   def self.bulk_update(ary)
+    updates_by_version = {}
+    updates_by_gem = {}
+
     ary.each do |full_name, count|
-      increment(full_name, count: count)
+      if updates_by_version.key?(full_name)
+        version, old_count = updates_by_version[full_name]
+        updates_by_version[full_name] = [version, old_count + count]
+      else
+        version = Version.find_by(full_name: full_name)
+        updates_by_version[full_name] = [version, count] if version
+      end
+    end
+
+    updates_by_version.values.each do |version, version_count|
+      updates_by_gem[version.rubygem_id] ||= 0
+      updates_by_gem[version.rubygem_id] += version_count
     end
+
```

Comment:
I wonder if it would make sense to wrap the code from line 69 to 82 in a database transaction.
I've seen cases where one of these (possibly many) updates would fail half way through and you wouldn't know if it was safe to retry the entire method call. (since you could be writing the same counts again after a retry).